src/provider/controller/ExpiryController.ts

```diff
@@ -0,0 +1,112 @@
+import {injectable, inject} from 'inversify';
+
+import {Store} from '../store/Store';
+import {Inject} from '../common/Injectables';
+import {StoredNotification} from '../model/StoredNotification';
+import {RemoveNotifications, RootAction, CreateNotification, ExpireNotification} from '../store/Actions';
+import {RootState} from '../store/State';
+
+import {AsyncInit} from './AsyncInit';
+
+@injectable()
+export class ExpiryController extends AsyncInit {
+    private readonly _store: Store<RootState, RootAction>;
+
+    private _nextExpiry: {
+        note: StoredNotification,
+        timerHandler: number
+    } | null = null;
+
+    public constructor(@inject(Inject.STORE) store: Store<RootState, RootAction>) {
+        super();
+
+        this._store = store;
+        this._store.onAction.add(this.onAction, this);
+    }
+
+    protected async init(): Promise<void> {
+        await this._store.initialized;
+
+        this.scheduleEarliestExpiry(Date.now());
+    }
+
+    private async onAction(action: RootAction): Promise<void> {
+        if (action instanceof CreateNotification) {
+            this.addNotification(action.notification);
+        } else if (action instanceof RemoveNotifications) {
+            this.removeNotifications(action.notifications);
+        }
+    }
+
+    private addNotification(note: StoredNotification): void {
+        if (!this._nextExpiry || this.expiresBefore(note, this._nextExpiry.note)) {
```

Comment:
I would populate a local ordered container from the store within this class at instantiation and do the book-keeping locally throughout module's lifetime to avoid swap every time and keep the store queries at a minimum, then your `scheduleEarliestExpiry` would be free.

Reply:
✅